bundler/lib/bundler/definition.rb

```diff
@@ -633,6 +637,12 @@ def materialize(dependencies)
       specs
     end
 
+    def reresolve_without(incomplete_specs)
+      sources.remote!
```

Comment:
should it be possible to re-resolve locally in the missing deps case, because we have the full gemwspec with the correct deps locally?

Reply:
Yes, it would be possible, but unfortunately, it would affect performance. Basically, we could apply the following patch:

```diff
diff --git a/bundler/lib/bundler/lazy_specification.rb b/bundler/lib/bundler/lazy_specification.rb
index 5e8a70161d1..91ce1df4cb1 100644
--- a/bundler/lib/bundler/lazy_specification.rb
+++ b/bundler/lib/bundler/lazy_specification.rb
@@ -136,12 +136,8 @@ def __materialize__(candidates, fallback_to_non_installable: Bundler.frozen_bund
       end
       if search.nil? && fallback_to_non_installable
         search = candidates.last
-      elsif search
-        if search.is_a?(StubSpecification)
-          search.dependencies = dependencies
-        elsif !source.is_a?(Source::Path) && search.runtime_dependencies.sort != dependencies.sort
-          raise IncorrectLockfileDependencies.new(self)
-        end
+      elsif search && !source.is_a?(Source::Path) && search.runtime_dependencies.sort != dependencies.sort
+        raise IncorrectLockfileDependencies.new(self)
       end
       search
     end
```

on top of what we have now and get the lockfile automatically fixed in that case too, but that would mean loading all installed specifications on every materialization, noticeably affecting performance.

So, as of now, we only do this when we have resolved remotely.

That said, there's no reason to force `sources.remote!` here, they should already be in remote mode, so I'll remove that line.

Reply:
I added a note about this at 9444599a68efae1e2cb1e8b3e83de5943ad46842, and also removed the unnecessary call to `sources.remote!`.